driver-core/src/main/com/mongodb/internal/logging/StructuredLogger.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.logging;
+
+import com.mongodb.connection.ClusterId;
+import com.mongodb.internal.VisibleForTesting;
+import com.mongodb.internal.diagnostics.logging.Logger;
+import com.mongodb.internal.diagnostics.logging.Loggers;
+import com.mongodb.internal.logging.StructuredLogMessage.Entry;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.mongodb.internal.VisibleForTesting.AccessModifier.PRIVATE;
+import static java.lang.String.format;
+
+/**
+ * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ */
+public class StructuredLogger {
+
+    private final Logger logger;
+
+    private static final Map<String, StructuredLoggingInterceptor> INTERCEPTORS = new HashMap<>();
```

Comment:
Are there any concerns about the behavior in production scenarios? (I don't think so given that `INTERCEPTORS` is a final variable.)  If that's the case, how worried should we be about the test scenario?  Are we willing to give up some performance here in production to make test scenarios safe?  I'm not so sure.

Reply:
> Are there any concerns about the behavior in production scenarios?

If we are choosing to write code (even test code) that we can't seem to show to be correct for the sake of not harming performance of production code, I think we ought to at the very least (necessary but not sufficient) demonstrate that the alternative code that we can show to be correct, measurably hurts production performance.

> Are we willing to give up some performance here in production to make test scenarios safe?

In a situation when there are no writes at all, which is what we have in production, OpenJDK JDK `CHM.get` is different from `HM.get` in the two following ways:

1.  it computes the key has eagerly, which is weird because algorithmically this is not needed, i.e., it's likely an oversight, but it is also likely that JIT optimizes that computation away and that's why the authors don't care;
2. it does a single volatile read of a shared variable instead of a plain read, which, when there are no writes to the variable, barely have overhead over plain reads; but even if there were writes, we just don't care about the cost of a volatile read at the very least because `ch.qos.logback.classic.Logger.isDebugEnabled` does at least one such read as a result of calling `CopyOnWriteArrayList.size()`.

Hence the question: what performance penalty are we talking about?